client/reader/tag-stream/header.jsx

```diff
@@ -5,77 +5,106 @@ import React from 'react';
 import classnames from 'classnames';
 import { localize } from 'i18n-calypso';
 import { connect } from 'react-redux';
+import { sample } from 'lodash';
 
 /**
  * Internal dependencies
  */
 import FollowButton from 'components/follow-button/button';
 import QueryReaderTagImages from 'components/data/query-reader-tag-images';
-import { getFirstImageForTag } from 'state/reader/tags/images/selectors';
+import { getTagImages } from 'state/reader/tags/images/selectors';
 import resizeImageUrl from 'lib/resize-image-url';
 import cssSafeUrl from 'lib/css-safe-url';
 import { decodeEntities } from 'lib/formatting';
 import Gridicon from 'components/gridicon';
 
 const TAG_HEADER_WIDTH = 830;
+const TAG_HEADER_HEIGHT = 140;
 
-const TagStreamHeader = (
-		{ tag, tagImage, isPlaceholder, showFollow, following, onFollowToggle, translate, hasBackButton }
-	) => {
-	const classes = classnames( {
-		'tag-stream__header': true,
-		'is-placeholder': isPlaceholder,
-		'has-back-button': hasBackButton
-	} );
-	const imageStyle = {};
+class TagStreamHeader extends React.Component {
 
-	if ( tagImage ) {
-		const imageUrl = resizeImageUrl( tagImage.url, { w: TAG_HEADER_WIDTH } );
-		const safeCssUrl = cssSafeUrl( imageUrl );
-		imageStyle.backgroundImage = 'url(//' + safeCssUrl + ')';
+	static propTypes = {
+		isPlaceholder: React.PropTypes.bool,
+		tag: React.PropTypes.string,
+		showFollow: React.PropTypes.bool,
+		following: React.PropTypes.bool,
+		onFollowToggle: React.PropTypes.func,
+		tagImages: React.PropTypes.array,
+		hasBackButton: React.PropTypes.bool
+	};
+
+	constructor( props ) {
+		super( props );
+		this.state = {
+			tag: props.tag,
+			tagImages: props.tagImages
+		};
+		this.pickNewTagImage( props );
 	}
 
-	return (
-		<div className={ classes }>
-			<QueryReaderTagImages tag={ tag } />
-			{ showFollow &&
-				<div className="tag-stream__header-follow">
-					<FollowButton
-						followLabel={ translate( 'Follow Tag' ) }
-						followingLabel={ translate( 'Following Tag' ) }
-						iconSize={ 24 }
-						following={ following }
-						onFollowToggle={ onFollowToggle } />
-				</div>
-			}
+	componentWillReceiveProps( nextProps ) {
+		this.pickNewTagImage( nextProps );
+	}
+
+	pickNewTagImage = ( props = this.props ) => {
+		if ( this.state && this.state.tagImages !== props.tagImages ) {
+			this.state = {
```

Comment:
this _shouldn't_ work, the only place you can set `this.state` safely is in the constructor. I'd rejigger this to have pickNewImage return the chosen one and have cwrp call it and then call setState if the chosen one changed. pickNewTagImage would also have to return `null` if no `tagImages` are passed in...